pkg/keystorage/keystorage.go

```diff
@@ -0,0 +1,193 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Package keystorage provides the key storage implementation.
+package keystorage
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"crypto/subtle"
+	"fmt"
+	"sync"
+
+	"github.com/ProtonMail/gopenpgp/v2/helper"
+
+	"github.com/cosi-project/runtime/api/key_storage"
+)
+
+// KeyStorage is a key storage that can be used to store and retrieve the master key.
+//
+//nolint:govet
+type KeyStorage struct {
+	mx         sync.Mutex
+	underlying key_storage.Storage
+}
+
+// SetMasterKey sets the master key for the key storage, encrypts it using public key and stores it in slot id.
+func (ks *KeyStorage) SetMasterKey(masterKey []byte, slotID, slotPublicKey string) error {
+	switch {
+	case len(masterKey) != 32:
+		return fmt.Errorf("master key can only be 32 bytes long")
+	case slotID == "":
+		return fmt.Errorf("slot id cannot be empty")
+	case slotPublicKey == "":
+		return fmt.Errorf("slot public key cannot be empty")
+	}
+
+	ks.mx.Lock()
+	defer ks.mx.Unlock()
+
+	if !isZero(&ks.underlying) {
+		return fmt.Errorf("key storage is already initialized")
+	}
+
+	encryptedSlot, err := helper.EncryptBinaryMessageArmored(slotPublicKey, masterKey)
+	if err != nil {
+		return fmt.Errorf("failed to encrypt slot '%s': %w", slotID, err)
+	}
+
+	ks.underlying.StorageVersion = key_storage.StorageVersion_STORAGE_VERSION_1
+	ks.underlying.KeySlot = map[string]*key_storage.KeySlot{
+		slotID: {
+			Algorithm:    key_storage.Algorithm_PGP_AES_GCM_256,
+			EncryptedKey: []byte(encryptedSlot),
+		},
+	}
+	ks.underlying.KeysHmacHash = ks.hashSlots(masterKey)
+
+	return nil
+}
+
+// SetKey sets the new key slot with the given slot id and public key using previus slot and its private key.
+// It cannot be used to update an existing key slot.
+// It is required to call SetMasterKey() or UnmarshalBinary() to initialize the key storage first.
+func (ks *KeyStorage) SetKey(newSlotID, newSlotPublicKey, oldSlotID, oldSlotPrivateKey string) error {
```

Comment:
we probably need `DeleteSlot` function as well

and not sure about proper naming here, `EncryptSlot` maybe?

Reply:
Changed to `AddKeySlot` and added `DeleteKeySlot`